pkg/orchestrator/controller/controller.go

```diff
@@ -0,0 +1,264 @@
+package controller
+
+import (
+	"flag"
+	"fmt"
+	"sync"
+
+	"github.com/submariner-io/admiral/pkg/federate"
+	submarinerClientset "github.com/submariner-io/submariner/pkg/client/clientset/versioned"
+	submarinerInformers "github.com/submariner-io/submariner/pkg/client/informers/externalversions"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+type clusterWatcherMap map[string]*clusterWatcher
+
+type clusterWatcher struct {
+	clusterID         string
+	stopChan          chan struct{}
+	clusterWorkQueue  workqueue.RateLimitingInterface
+	endpointWorkQueue workqueue.RateLimitingInterface
+}
+
+type controller struct {
+	sync.Mutex
+	federator       federate.Federator
+	clusterWatchers clusterWatcherMap
+
+	// Indirection hook for unit tests to supply fake client sets
+	newSubmClientset func(kubeConfig *rest.Config) (submarinerClientset.Interface, error)
+}
+
+// clusterIDLabelKey is the key for a label that is added to the federated submariner Cluster and Endpoint resources to
+// hold the ID of the cluster from which the resource originated, allowing us to only process watch notifications
+// emanating from the originating cluster.
+const clusterIDLabelKey = "submariner-io/clusterID"
+
+var (
+	submarinerNamespace string
+)
+
+func init() {
+	flag.StringVar(&submarinerNamespace, "submariner-namespace", "submariner", "The namespace in which the submariner components are deployed.")
+}
+
+func New(federator federate.Federator) *controller {
+	return &controller{
+		federator:       federator,
+		clusterWatchers: make(clusterWatcherMap),
+
+		newSubmClientset: func(c *rest.Config) (submarinerClientset.Interface, error) {
+			return submarinerClientset.NewForConfig(c)
+		},
+	}
+}
+
+func (c *controller) Start() error {
+	err := c.federator.WatchClusters(c)
+	if err != nil {
+		return fmt.Errorf("Could not register cluster watch: %v", err)
+	}
+
+	return nil
+}
+
+func (c *controller) Stop() {
+	c.Lock()
+	defer c.Unlock()
+
+	for _, clusterWatcher := range c.clusterWatchers {
+		clusterWatcher.close()
+	}
+}
+
+func (c *controller) removeExistingClusterWatcher(clusterID string) {
```

Comment:
Perhaps it would be more correct to call this only n the context of update/remove since on add you don't expect it to actually exist already.